raise exc when we fail to create a container

```
self = <sanity.TestContainer object at 0x7f225f487be0>

    def test_ethtool(self):
        networks_devices = os.listdir("/sys/class/net")
        for device in networks_devices:
>           self.container.execute(["ethtool", device])

tests/sanity.py:40:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/root/.local/lib/python3.6/site-packages/conu/backend/docker/container.py:247: in execute
    exec_i = self.d.exec_create(self.get_id(), command, **exec_create_kwargs)
/usr/lib/python3.6/site-packages/docker/utils/decorators.py:34: in wrapper
    return f(self, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <docker.api.client.APIClient object at 0x7f225d50da20>, resource_id = '', args = (['ethtool', 'wlp3s0'],), kwargs = {}

    @functools.wraps(f)
    def wrapped(self, resource_id=None, *args, **kwargs):
        if resource_id is None and kwargs.get(resource_name):
            resource_id = kwargs.pop(resource_name)
        if isinstance(resource_id, dict):
            resource_id = resource_id.get('Id', resource_id.get('ID'))
        if not resource_id:
            raise errors.NullResource(
>               'Resource ID was not provided'
            )
E           docker.errors.NullResource: Resource ID was not provided

/usr/lib/python3.6/site-packages/docker/utils/decorators.py:17: NullResource
--------------------------------------------------------------------------- Captured stderr setup ----------------------------------------------------------------------------
/usr/bin/docker-current: --pid: invalid PID mode.
See '/usr/bin/docker-current run --help'.
```

--- a/conu/backend/docker/container.py
+++ b/conu/backend/docker/container.py
@@ -259,16 +259,27 @@ class DockerContainer(Container):
         run_command_instance.image_name = image.get_id()
         run_command_instance.options += ["-d"]
         popen_instance = subprocess.Popen(run_command_instance.build(), stdout=subprocess.PIPE)
-        container_id = popen_instance.communicate()[0].strip().decode("utf-8")
-        return cls(image, container_id)
+        stdout = popen_instance.communicate()[0].strip().decode("utf-8")
+        if popen_instance.returncode > 0:
+            raise ConuException("Container exited with an error: %s" % popen_instance.returncode)
+        # no error, stdout is the container id
+        return cls(image, stdout)
 
     @classmethod
     def run_via_binary_in_foreground(
             cls, image, run_command_instance=None, popen_params=None, container_name=None):
         """
-        create container using provided image and run it in foreground;
+        Create container using provided image and run it in foreground;
         this method is useful to test real user scenarios when users invoke containers using
-        binary and pass input into the container via STDIN
+        binary and pass input into the container via STDIN. Please bear in mind that conu doesn't
+        know the ID of the container when created like this, so it's highly recommended to name
+        your container. You are also responsible for checking whether the container exited
+        successfully via:
+
+            container.popen_instance.returncode
+
+        Please consult the documentation for subprocess python module for best practices on
+        how you should work with instance of Popen
 
         :param image: instance of Image
         :param run_command_instance: instance of DockerRunCommand

--- a/tests/integration/test_docker.py
+++ b/tests/integration/test_docker.py
@@ -5,10 +5,12 @@ import os
 import subprocess
 import time
 
+import pytest
+
 from .constants import FEDORA_MINIMAL_REPOSITORY, FEDORA_MINIMAL_REPOSITORY_TAG, THE_HELPER_IMAGE, \
     FEDORA_REPOSITORY
 
-from conu import DockerContainer, DockerImage, DockerRunCommand, Probe
+from conu import DockerContainer, DockerImage, DockerRunCommand, Probe, ConuException
 
 from six import string_types
 
@@ -85,6 +87,22 @@ def test_copy_from(tmpdir):
         c.delete()
 
 
+def test_container_create_failed():
+    image = DockerImage(FEDORA_MINIMAL_REPOSITORY, tag=FEDORA_MINIMAL_REPOSITORY_TAG)
+    # should raise an exc, there is no such command: waldo; we need to find waldo first
+    with pytest.raises(ConuException):
+        DockerContainer.run_via_binary(
+            image,
+            DockerRunCommand(command=["waldo"])
+        )
+    c = DockerContainer.run_via_binary_in_foreground(
+        image,
+        DockerRunCommand(command=["waldo"])
+    )
+    c.popen_instance.communicate()
+    assert c.popen_instance.returncode > 0
+
+
 def test_networking_scenario():
     """
     Listen via netcat in one container, send a secret message to the container via another one.
@@ -152,6 +170,7 @@ def test_wait_for_status():
     assert end > 2, "Probe should wait till container status is exited"
     assert end < 7, "Probe should end when container status is exited"
 
+
 def test_exit_code():
     image = DockerImage(FEDORA_MINIMAL_REPOSITORY, tag=FEDORA_MINIMAL_REPOSITORY_TAG)
     cmd = DockerRunCommand(command=['sleep', '2'])
@@ -161,6 +180,7 @@ def test_exit_code():
     p.run()
 
     assert not cont.is_running() and cont.exit_code() == 0
-    cmd = DockerRunCommand(command=['bubububu'])
+    cmd = DockerRunCommand(command=['bash', '-c', "exit 42"])
     cont = DockerContainer.run_via_binary(image, cmd)
-    assert cont.exit_code() == 127
+    cont.wait()
+    assert cont.exit_code() == 42

